fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAccrualActivityProcessingServiceImpl.java

```diff
@@ -92,42 +92,94 @@ public void processAccrualActivityForLoanClosure(@NotNull Loan loan) {
+            if (!installment.isDownPayment()) { // Exclude downpayment installments
+                feeChargesPortion = MathUtil.add(feeChargesPortion, installment.getFeeChargesCharged());
+                penaltyChargesPortion = MathUtil.add(penaltyChargesPortion, installment.getPenaltyCharges());
+                interestPortion = MathUtil.add(interestPortion, installment.getInterestCharged());
+            }
         }
+
+        // Subtract already posted accrual activities
         List<LoanTransaction> accrualActivities = loan.getLoanTransactions(t -> t.isAccrualActivity() && !t.isReversed());
-        // subtract already posted activities
         for (LoanTransaction accrualActivity : accrualActivities) {
-            if (MathUtil.isLessThan(feeChargesPortion, accrualActivity.getFeeChargesPortion())
-                    || MathUtil.isLessThan(penaltyChargesPortion, accrualActivity.getPenaltyChargesPortion())
-                    || MathUtil.isLessThan(interestPortion, accrualActivity.getInterestPortion())) {
-                reverseAccrualActivityTransaction(accrualActivity);
-            } else {
-                feeChargesPortion = MathUtil.subtract(feeChargesPortion, accrualActivity.getFeeChargesPortion());
-                penaltyChargesPortion = MathUtil.subtract(penaltyChargesPortion, accrualActivity.getPenaltyChargesPortion());
-                interestPortion = MathUtil.subtract(interestPortion, accrualActivity.getInterestPortion());
+            feeChargesPortion = MathUtil.subtract(feeChargesPortion, accrualActivity.getFeeChargesPortion());
+            penaltyChargesPortion = MathUtil.subtract(penaltyChargesPortion, accrualActivity.getPenaltyChargesPortion());
+            interestPortion = MathUtil.subtract(interestPortion, accrualActivity.getInterestPortion());
+        }
+
+        // Check each past installment for accrual activity
+        for (LoanRepaymentScheduleInstallment installment : loan.getRepaymentScheduleInstallments()) {
+            if (!installment.isDownPayment()) {
+                List<LoanTransaction> installmentAccruals = accrualActivities.stream()
+                        .filter(t -> t.getDateOf().isEqual(installment.getDueDate())).toList();
+
+                if (installmentAccruals.isEmpty()) {
+                    // No AAT for this installment; create one
+                    createAccrualActivityTransaction(loan, installment, transactionDate);
+
+                    // Subtract processed portions
+                    feeChargesPortion = MathUtil.subtract(feeChargesPortion, installment.getFeeChargesCharged());
+                    penaltyChargesPortion = MathUtil.subtract(penaltyChargesPortion, installment.getPenaltyCharges());
+                    interestPortion = MathUtil.subtract(interestPortion, installment.getInterestCharged());
+                } else if (installmentAccruals.size() > 1 || !isAccrualActivityConsistent(installment, installmentAccruals.get(0))) {
```

Comment:
For the above reasons i dont think the 1st part of the condition would ever be true.